File: cprdspy/CPR_plotly1/Waves/waves.py
import plotly.graph_objects as go
import numpy as np
from ..Circles.circle import concentric_circles


def wave_ari(
    A=1,
    F=3,
    P=12,
    rotation=0,
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制等差波形。
    """
    traces = []
    for i in range(P + 1):
        circle_traces = concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rotation),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rotation),
            ),
            n=F,
            radius=R,
            param=A,
            mode="arithmetic",
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            line_width=line_width,
            show_center=show_center,
            **kwargs,
        )
        traces.extend(circle_traces)
    return traces


def wave_geo(
    A=1,
    F=3,
    P=12,
    rotation=0,
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制等比波形。
    """
    traces = []
    for i in range(P + 1):
        circle_traces = concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rotation),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rotation),
            ),
            n=F,
            radius=R,
            param=A,
            mode="geometric",
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            line_width=line_width,
            show_center=show_center,
            **kwargs,
        )
        traces.extend(circle_traces)
    return traces


def wave(
    A=1,
    F=3,
    P=12,
    rotation=0,
    mode="geometric",
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制波形（通用）。
    """
    traces = []
    for i in range(P + 1):
        circle_traces = concentric_circles(
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rotation),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rotation),
            ),
            n=F,
            radius=R,
            param=A,
            mode=mode,
            direction=direction,
            color=color,
            alpha=alpha,
            points=points,
            line_width=line_width,
            show_center=show_center,
            **kwargs,
        )
        traces.extend(circle_traces)
    return traces


def wave_wave(
    A=1,
    F=3,
    P=12,
    rotation=0,
    mode="geometric",
    direction="both",
    color="#0f0",
    alpha=1,
    R=1,
    center=(0, 0),
    points=1000,
    line_width=1,
    show_center=True,
    **kwargs,
):
    """
    绘制嵌套波形。
    """
    traces = []
    for i in range(P + 1):
        wave_traces = wave(
            A=A,
            F=F,
            P=P,
            rotation=rotation,
            mode=mode,
            direction=direction,
            color=color,
            alpha=alpha,
            R=R,
            center=(
                center[0] + np.cos(i * 2 * np.pi / P + np.pi / 2 + rotation),
                center[1] + np.sin(i * 2 * np.pi / P + np.pi / 2 + rotation),
            ),
            points=points,
            line_width=line_width,
            show_center=show_center,
            **kwargs,
        )
        traces.extend(wave_traces)
    return traces

File: cprdspy/CPR_js2/Waves/waves.js
/**
 * CPR 图形库 - JavaScript 版本 (Plotly.js)
 * Waves 模块 - 波浪图案相关函数
 */

import { concentricCircles } from "../Circles/circle.js";

/**
 * 绘制等差波浪
 */
function waveAri({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    for (let i = 0; i <= P; i++) {
        const newCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circleTraces = concentricCircles({
            center: newCenter,
            n: F,
            radius: R,
            param: A,
            mode: "arithmetic",
            direction,
            color,
            alpha,
            points,
            lineWidth,
            showCenter,
        });

        traces.push(...circleTraces);
    }

    return traces;
}

/**
 * 绘制等比波浪
 */
function waveGeo({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    for (let i = 0; i <= P; i++) {
        const newCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circleTraces = concentricCircles({
            center: newCenter,
            n: F,
            radius: R,
            param: A,
            mode: "geometric",
            direction,
            color,
            alpha,
            points,
            lineWidth,
            showCenter,
        });

        traces.push(...circleTraces);
    }

    return traces;
}

/**
 * 绘制波浪
 */
function wave({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    mode = "geometric",
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    for (let i = 0; i <= P; i++) {
        const newCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const circleTraces = concentricCircles({
            center: newCenter,
            n: F,
            radius: R,
            param: A,
            mode,
            direction,
            color,
            alpha,
            points,
            lineWidth,
            showCenter,
        });

        traces.push(...circleTraces);
    }

    return traces;
}

/**
 * 绘制嵌套波浪
 */
function waveWave({
    A = 1,
    F = 3,
    P = 12,
    rotation = 0,
    mode = "geometric",
    direction = "both",
    color = "#0f0",
    alpha = 1,
    R = 1,
    center = [0, 0],
    points = 1000,
    lineWidth = 1,
    showCenter = true,
} = {}) {
    const traces = [];

    for (let i = 0; i <= P; i++) {
        const newCenter = [
            center[0] + Math.cos((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
            center[1] + Math.sin((i * 2 * Math.PI) / P + Math.PI / 2 + rotation),
        ];

        const waveTraces = wave({
            A,
            F,
            P,
            rotation,
            mode,
            direction,
            color,
            alpha,
            R,
            center: newCenter,
            points,
            lineWidth,
            showCenter,
        });

        traces.push(...waveTraces);
    }

    return traces;
}

export { waveAri, waveGeo, wave, waveWave };
